### DISCUSSION TITLE
docs: How to get value of frequencies?

### DISCUSSION THREAD

**Discussion by BigGitWorld** [alnitak/flutter_soloud#21] [General]
Hello.
I want to plot a Line chart from frequency domain of a wav file (x-axis: Frequencies, y-axis: decibel values).
How can I get value of frequencies?

**Comment by alnitak** [comment]
Hi,
you can only get FFT and  wave data while the audio is playing. See *getFft()*, *getWave()*, *getAudioTexture()* in the doc.
They return a convenient `Pointer<Float>` for performance instead of copying it ie in a Float32List every time.

Getting a whole audio frequency domain at once is not yet planned

**Comment by BigGitWorld** [comment]
> Hi, you can only get FFT and wave data while the audio is playing. See _getFft()_, _getWave()_, _getAudioTexture()_ in the doc. They return a convenient `Pointer<Float>` for performance instead of copying it ie in a Float32List every time.
> 
> Getting a whole audio frequency domain at once is not yet planned

Thanks for your answer.
1. But I didn't understand how can I get value of frequencies to Hz from FFT data. Do you mean `ffi.Pointer<ffi.Float> audioData` represents amplitude of each sine wave or frequency of sine wave? I want to have both of them (a Map with key=frequency and value=amplitude of that sine waves that create our signal). How can I find it?
could you give me source code of [this example](https://raw.githubusercontent.com/alnitak/flutter_soloud/main/img/audacity_spectrum.png) where x-axis shows frequencies and y-axis shows amplitude of them to decibel?

3. In fact, I have a reference wav file and want to compare other wav files with it to check if they are similar or not.
can you suggest me a solution please?

**Comment by alnitak** [comment]
1.
I notice now that I didn't exposed *getAudioTexture()*, it's not available on Flutter side. So you will need to use *getAudioTexture2D()*.

Using *getAudioTexture2D()* you will get back a matrix with 256 rows of 512 floats. You could consider only the 1st row as it represents the latest data. So this first row is composed by:
- The first 256 floats represent the FFT frequencies data ~~in the 0.0~1.0 range~~ > 0.0.
- The other 256 floats represent the wave data in -1.0~1.0 range (amplitude).
So, the first 256 floats in this array, I think are the values you are looking for:
```
final ffi.Pointer<ffi.Float> audioData = SoLoud().getAudioTexture2D(playerData);
```

You can access these floats with `double n = audioData[i]`. The `i` index is the frequency in Hz and its value is the amplitude (volume) of that frequency.
Since the engine is set to 44100 samples/s and the amount of data given to compute FFT is 1024, below a function to know which frequency is represented in the array index `i`.

```
double Index2Freq(int i, double samples, int nFFT) {
  return  (samples / nFFT / 2.0) * i;
}
```
and you can use it with something like this:
```
void getFreqs() {
  for (final int i=0; i<256; i++) print('index: $i  is  ${Index2Freq(i, 44100, 1024)} Hz');
}
```
[Here](https://github.com/alnitak/flutter_soloud/blob/07b86999ecda21379ac07773d6521d5497b60da6/example/lib/visualizer/bars_fft_widget.dart#L70C6-L70C6) you can see the widget used in the `Visualizer` example to show the FFT data, but the correct image is [this](https://github.com/alnitak/flutter_soloud/raw/main/img/flutter_soloud_spectrum.png). The one you linked is made using Audacity app and it's a bit different with the one generated by this plugin and I don't know yet why :smile: 

I didn't tested and I don't know much about audio processing, but hope this helps!

2. I think this could be more complicated than it might seems. To compare audio you should use *fingerprints*, basically you should generate these fingerprint of samples and then compare each other. There are some libraries around by I never tried.

I move this issue to Discussion, it maybe be useful to others!

**Comment by BigGitWorld** [comment]
1. why the The first 256 floats  that represents ``FFT frequencies data`` is in range of (0.0~1.0)?
2. why the second 256 floats that represents ``wave data`` is in range of (-1.0~1.0)?

**Comment by BigGitWorld** [comment]
No reply?

**Comment by alnitak** [comment]
Oh! Sorry, I missed it!
By default, float values are used for faster interaction with audio data. It's possible, to modify the plugin code (not exposed to Flutter) and use integers, but I have chosen to stick with float numbers.

Wave data is in the -1~1 range because I choose to use `Soloud::CLIP_ROUNDOFF`. [Here](https://solhsa.com/soloud/soloud_20200207.html#clipping) some Soloud docs about clipping.
FFT data has values >0 (I have edited the previous post). This means that values can be greater than 1.0 and it is due to the windowing method used. Here I used [Blackman windowing](https://github.com/alnitak/flutter_soloud/blob/fe0cb2babfcacd11ceaa265cbc0877f7e07fd534/src/analyzer.cpp#L56) which is used also by ShaderToy, even if I'm not fully satisfied.
